bin/picom-trans

```diff
@@ -46,194 +54,306 @@ print_usage() {
-      ;;
-      r) action='reset' ;;
-      t) action='toggle' ;;
-      d) action='delete' ;;
-      g) action='get' ;;
-      n) wprefix='-name'; window=$OPTARG ;;
-      w) wprefix='-id'; window=$OPTARG ;;
-      o) opacity=$OPTARG ;;
-      \?) break ;;
+    # Separate  "--ARG=VAL"  into  "--ARG" "VAL"
+    case "$1" in
+      (--opacity=*|--name=*|--window=*)
+        ARG="$(echo "$1" | sed -E 's/(--[^=]+)=.*$/\1/')"
+        VAL="${1##${ARG}=}"
+        shift && set -- "$ARG" "$VAL" "$@"
+    esac
+
+    # Turn into short form
+    case "$1" in
+      (--help|--opacity|--get|--delete|--toggle|--reset|--select|--current|--name|--window)
+        ARG=${1#-}                      # remove one '-' from prefix
+        ARG="$(echo $ARG | cut -c -2)"  # get first two characters
+        shift && set -- "$ARG" "$@"
     esac
+
+    # If the argument still starts with --, it is an invalid argument
+    case "$1" in
+      (--*)
+        echo "$EXE_NAME: illegal option $1" >&2
+        exit 1
+    esac
+
+    # Prepare for next iteration
+    ARG="$1"
+    shift && set -- "$@" "$ARG"
+    i=$(( i + 1 ))
   done
 
-  # Read positional arguments
-  shift $((OPTIND - 1))
-  test -n "$1" && opacity=$1 && shift
-done
 
-# clean up opacity. xargs == a poor man's trim.
-opacity=$(echo "$opacity" | xargs | sed 's/%//g')
+  while test $# -gt 0; do
+    # 15 +15 -15 are opacity values
+    # filter them out here, otherwise getopts will complain
+    if echo "$1" | grep -qE '^[+-]?[[:digit:]]+%?$'; then
+      target_opacity=$1
+      shift
+      continue
+    fi
 
-# Validate opacity value
-if test -z "$action" && ! echo "$opacity" | grep -qE '^[+-]?[0-9]+$'; then
-  echo "Invalid opacity specified: $opacity."
-  exit 1
-fi
+    OPTIND=1
+    while getopts ':ho:gdtrscn:w:' OPTION
+    do
+      case "$OPTION" in
+        (h) print_usage; exit 0;;
+        (o) target_opacity="$OPTARG";;
+        (g) action=get;;
+        (d) action=delete;;
+        (t) action=toggle;;
+        (r) action=reset;;
+        (s) winidtype=;         winid=;;
+        (c) winidtype=-id;      winid="$(get_focused_window_id)";;
+        (n) winidtype=-name;    winid="$OPTARG";;
+        (w) winidtype=-id;      winid="$OPTARG";;
+        (\?)
+          # This might not be an invalid argument.
+          # It could be an opacity value.
+          #
+          # If it is an opacity value,  just break.
+          # It will be handled at next iteration of the loop.
+          if echo $OPTARG | grep -q '[[:digit:]]'
+          then break
+          else echo "$EXE_NAME: illegal option: $1" >&2 && exit 1
+          fi
+      esac
 
-# Reset opacity for all windows
-if test x"$action" = x'reset'; then
-  xwininfo -root -tree \
-  | sed -n 's/^     \(0x[[:xdigit:]]*\).*/\1/p' \
-  | while IFS=$(printf '\n') read wid; do
-    xprop -id "$wid" -remove _NET_WM_WINDOW_OPACITY
+      # If getopts goes to the second argument, stop
+      test $OPTIND -ge 3 && break
+    done
+    shift
```

Comment:
Specifying the relative opacity (i.e. `-10`) now work correctly without `-o`.
But out of curiosity, does anything speak against calling `getopts` for each argument separately (like the comment actually implies)?
```diff
--- a/bin/picom-trans
+++ b/bin/picom-trans
@@ -125,22 +125,13 @@ function parse_args
         (c) winidtype=-id;      winid="$(get_focused_window_id)";;
         (n) winidtype=-name;    winid="$OPTARG";;
         (w) winidtype=-id;      winid="$OPTARG";;
-        (\?)
-          # This might not be an invalid argument.
-          # It could be an opacity value.
-          #
-          # If it is an opacity value,  just break.
-          # It will be handled at next iteration of the loop.
-          if echo $OPTARG | grep -q '[[:digit:]]'
-          then break
-          else echo "$EXE_NAME: illegal option: $1" >&2 && exit 1
-          fi
+        (\?) echo "$EXE_NAME: illegal option: $1" >&2 && exit 1;;
       esac
 
-      # If getopts goes to the second argument, stop
-      test $OPTIND -ge 3 && break
+      # If getopts goes to the second argument, stop and check for an opacity value first
+      test $OPTIND -ge 2 && OPTIND=$(( $OPTIND - 1 )) && break
     done
-    shift
+    shift $OPTIND
   done
 } #}
 
```
This works fine for me with `picom-trans -w WINID -50%` and `picom-trans -c -50%`.

Reply:
Let's test. Add the following lines -
```diff
@@ -115,6 +115,7 @@ function parse_args
     OPTIND=1
     while getopts ':ho:gdtrscn:w:' OPTION
     do
+      echo $OPTIND:$OPTION:$OPTARG
       case "$OPTION" in
         (h) print_usage; exit 0;;
         (o) target_opacity="$OPTARG";;
@@ -143,6 +144,7 @@ function parse_args
     done
     shift
   done
+  exit 0
 } #}
```

---

Now, run the command `picom-trans -co80 -sg --name st`, once with your patch applied, once without your patch applied.

---

### With your patch applied
```
$ picom-trans -co80 -sg --name st
2:c:
2:s:
3:n:st
```
### Without your patch applied
```
$ picom-trans -co80 -sg --name st
2:c:
2:o:80
3:s:
2:s:
2:g:
4:n:st
3:n:st
```

---

Notice the missing arguments with you patch applied?

Also, please note that the current detection process isn't that great either. Some arguments, as you may have noticed, are being repeated (twice).

But, IMO, detecting arguments twice is better than not detecting them at all, isn't it? 😉

Reply:
### UPDATE:

Try the following patch, and run 
```
DEBUG=getopts picom-trans -co80 -sg --name st -nst
```
and see if the output makes any sense.

---

```diff
diff --git a/bin/picom-trans b/bin/picom-trans
index e9b8bbd..9435bf9 100755
--- a/bin/picom-trans
+++ b/bin/picom-trans
@@ -103,6 +103,13 @@ function parse_args
   done
 
 
+  # Feature to allow to easily test getopts parsing
+  if test "$DEBUG" = "getopts"
+  then function DEBUG_getopts { "$@"; }
+  else function DEBUG_getopts { :; }
+  fi
+
+
   while test $# -gt 0; do
     # 15 +15 -15 are opacity values
     # filter them out here, otherwise getopts will complain
@@ -115,6 +122,16 @@ function parse_args
     OPTIND=1
     while getopts ':ho:gdtrscn:w:' OPTION
     do
+
+      DEBUG_getopts echo
+      DEBUG_getopts echo "$*"
+      DEBUG_getopts echo $OPTIND:$OPTION:$OPTARG
+
+      # If getopts goes to the second argument, stop
+      test -z "$OPTARG" -a $OPTIND -ge 3 -o $OPTIND -ge 4 && break
+
+      DEBUG_getopts echo $OPTIND:$OPTION:$OPTARG
+
       case "$OPTION" in
         (h) print_usage; exit 0;;
         (o) target_opacity="$OPTARG";;
@@ -137,12 +154,10 @@ function parse_args
           else echo "$EXE_NAME: illegal option: $1" >&2 && exit 1
           fi
       esac
-
-      # If getopts goes to the second argument, stop
-      test $OPTIND -ge 3 && break
     done
     shift
   done
+  DEBUG_getopts exit 0
 } #}
```

Reply:
Seems like the above patch has fixed the problem of detecting an argument twice. 

Nice!

Reply:
I still get double arguments with the above patch. Either `getopts` is a bit weird or I don't get it…
I've tested with `picom-trans -ctgo80 -sg --name st -nst -ggst 20 --opacity 50 -30`,  for maximum confusion.

Can you test whether replacing the skipping `test … && break` with
```sh
      test -n "$OPTARG" -a $OPTIND -ge 3 && shift && break
      test $OPTIND -ge 2 && break
```
and placing it after the `case` block still skips arguments? It seems `shift`ing based on `$OPTIND` isn't the right way.

BTW: Using the current HEAD and/or patch results in the wrong argument being printed as *illegal option* (e.g. with `picom-trans -co80 -x`). This works as expected with moved `test … && break` and doesn't need the additional logic inside the `case` block.

Reply:
On current HEAD, my result of `picom-trans -co80 -sg --name st` is -
```
2:c:
2:o:80
3:s:
2:s:
2:g:
4:n:st
3:n:st
```

My `sh` is OpenBSD 6.9 `/bin/sh`, which is probably hard-linked to `/bin/ksh`, i.e. the Korn Shell.

Reply:
That's a pretty large difference between our outputs!

Reply:
I really don't know what to do now.... Should we file a bug report for Bash? Should we ask in the openbsd mailing list? 

Also, what should be the expected output according to POSIX?  
🤷

Reply:
Well, looks like bash is the culprit here.

This is the output of `picom-trans -co80 -sg --name st` in `dash` (debian almquist shell) -
```
2:c:
2:o:80
3:s:
2:s:
2:g:
4:n:st
3:n:st
```

Reply:
Here's `zsh` -
```
1:c:
2:o:80
2:s:
2:g:
5:n:st
1:s:
1:g:
4:n:st
3:n:st
```

And yes, it is different from `bash` and it is also different from `ksh`/`dash`

Reply:
🤢

Reply:
Following the discussion in [this mailing-list](https://www.mail-archive.com/austin-group-l%40opengroup.org/msg04112.html), `$OPTIND` is not well-(enough-)defined inside the `while getopts …; do` loop (at least for the implementations to exhibit the same behaviour). The only things defined in [the standard](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/getopts.html#tag_20_54_03) seem to be:
- Before processing, initialize to `OPTIND=1` (although some shells seem to have problems with internal state, but that's besides the point)
- After exiting with a non-zero status (i.e. finished argument processing) `$OPTIND` is set to the index after the last argument

I personally think the *bash-way* is closest to my intuitive understanding, while the `ksh`/`dash`/`busybox` behaviour seems to be the exact opposite. To understand `zsh` I probably have to look deeper into our outer loop.

However, all of that is completely irrelevant as we can't rely on "implementation-dependent" behaviour.

Reply:
Right. Thanks for the link to the mailing-list, and also for summarizing the main points here.

In ccf5f09, we don't check `$OPTIND` at all, and just do all of our processing before getopts is called. So, it should fix/workaround this issue.